confmap/provider/s3mapprovider/provider_test.go

```diff
@@ -0,0 +1,147 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3mapprovider
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/confmap"
+)
+
+// A s3 client mocking s3mapprovider works in normal cases
+type testClient struct{}
+
+// Implement GetObject() for testClient in normal cases
+func (client *testClient) GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	// read local config file and return
+	f, err := ioutil.ReadFile("./testdata/otel-config.yaml")
+	if err != nil {
+		return &s3.GetObjectOutput{}, err
+	}
+	return &s3.GetObjectOutput{Body: io.NopCloser(bytes.NewReader(f)), ContentLength: (int64)(len(f))}, nil
+}
+
+// Create a provider mocking s3mapprovider works in normal cases
+func NewTestProvider() confmap.Provider {
+	return &provider{client: &testClient{}}
+}
+
+// A s3 client mocking s3mapprovider works when the returned config file is invalid
+type testInvalidClient struct{}
+
+// Create a provider mocking s3mapprovider works when the returned config file is invalid
+func NewTestInvalidProvider() confmap.Provider {
+	return &provider{client: &testInvalidClient{}}
+}
+
+// Implement GetObject() for testClient when the returned config file is invalid
+func (client *testInvalidClient) GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	// read local config file and return
+	f := []byte("wrong yaml:[")
+	return &s3.GetObjectOutput{Body: io.NopCloser(bytes.NewReader(f)), ContentLength: (int64)(len(f))}, nil
+}
+
+// A s3 client mocking s3mapprovider works when there is no corresponding config file according to the given s3-uri
+type testNonExistClient struct{}
+
+// Create a provider mocking s3mapprovider works when there is no corresponding config file according to the given s3-uri
+func NewTestNonExistProvider() confmap.Provider {
+	return &provider{client: &testNonExistClient{}}
+}
+
+// Implement GetObject() for testClient when there is no corresponding config file according to the given s3-uri
+func (client *testNonExistClient) GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	// read local config file and return
+	f := []byte("wrong yaml:[")
+	return &s3.GetObjectOutput{Body: io.NopCloser(bytes.NewReader(f)), ContentLength: (int64)(len(f))}, nil
```

Comment:
I don't think this implementation is correct.  Shouldn't this return an error of some sort?

Reply:
Hi Anthony, before I implement like this because our code will then also check if the config file is valid or not.

```
        // read config from response body
	dec := yaml.NewDecoder(resp.Body)
	defer resp.Body.Close()
	var conf map[string]interface{}
	err = dec.Decode(&conf)
	if err != nil {
		return confmap.Retrieved{}, err
	}
	return confmap.NewRetrieved(conf)
```
Fixed and pushed!